### PR TITLE
Query history

### PR DESCRIPTION
**Pull request by qinghui-xu**
An implementation to persist query history to a relational database.

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/server/extension/query/history/QueryHistorySQLStore.java [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+        queryJsonParser.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
+        queryJsonParser.enableDefaultTyping();
+    }
+
+    public static ObjectMapper getQueryJsonParser()
+    {
+        return queryJsonParser;
+    }
+
+    private static final String CREATE_TABLE = "create table query_history (" +
```
Comment (szehon-ho) [inline presto-main/src/main/java/com/facebook/presto/server/extension/query/history/QueryHistorySQLStore.java]: It's not exposed right?  Maybe just put as a comment.

File: presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java [hunk 2]
```diff
@@ -163,8 +218,12 @@ public boolean addQuery(T execution)
      */
     public void expireQuery(QueryId queryId)
     {
-        tryGetQuery(queryId)
-                .ifPresent(expirationQueue::add);
+        Optional<T> expiredQuery = tryGetQuery(queryId);
+        expiredQuery.ifPresent(expirationQueue::add);
+        // Save the expired query to the history store with help of the extension.
+        expiredQuery.ifPresent(query ->
```
Comment (szehon-ho) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java:232]: Is this on the critical path?  (writing to MySQL).

File: presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java [hunk 3]
```diff
@@ -46,10 +52,13 @@
 import static java.util.Objects.requireNonNull;
 
 @ThreadSafe
-public class QueryTracker<T extends TrackedQuery>
+public class QueryTracker<T extends QueryExecution>
 {
     private static final Logger log = Logger.get(QueryTracker.class);
 
+    private static final String CONFIG_RESOURCE = "META-INF/presto/extension/history/jdbc.properties";
```
Comment (piyushnarang) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java]: might be a good idea to read this config similarly to the resource manager properties / session properties. 
We could add a todo to tackle this in the future

File: presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java [hunk 4]
```diff
@@ -46,10 +52,13 @@
...
+    public static final String QUERY_HISTORY_EXTENSION = "extension.query.history";
```
Comment (piyushnarang) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java]: if I understand correctly, this is really just used to indicate if the config resource is there or not. Can we drop this to just directly check if the resource is there / not? If the file doesn't exist skip?

File: presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java [hunk 5]
```diff
@@ -147,6 +186,22 @@ public T getQuery(QueryId queryId)
                 .orElseThrow(NoSuchElementException::new);
     }
 
+    public BasicQueryInfo getCurrentOrPastBasicQueryInfo(QueryId queryId)
+    {
+        return tryGetQuery(queryId).map(QueryExecution::getBasicQueryInfo)
+                .orElseGet(() ->
+                        queryHistoryStore.map(store -> store.getBasicQueryInfo(queryId))
+                                .orElseThrow(NoSuchElementException::new));
```
Comment (piyushnarang) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java:202]: should we throw? or return empty?
We could log in queryHistoryStore.getBasicQueryInfo / queryHistoryStore.getFullQueryInfo a warning if an attempt is made to retrieve a queryId that isn't found / can't be deserialized.

File: presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java [hunk 6]
```diff
@@ -147,6 +186,22 @@ public T getQuery(QueryId queryId)
...
+    }
+
+    public QueryInfo getCurrentOrPastFullQueryInfo(QueryId queryId)
+    {
+        return tryGetQuery(queryId).map(QueryExecution::getQueryInfo)
+                .orElseGet(() ->
+                        queryHistoryStore.map(store -> store.getFullQueryInfo(queryId))
+                                .orElseThrow(NoSuchElementException::new));
```
Comment (piyushnarang) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java:210]: here too? maybe return empty?

File: presto-main/src/main/java/com/facebook/presto/server/extension/query/history/QueryHistorySQLStore.java (hunk 1 above)
Comment (piyushnarang) [inline presto-main/src/main/java/com/facebook/presto/server/extension/query/history/QueryHistorySQLStore.java]: wdyt of following the structure that is currently used in the Presto code for reading / refreshing resource management configs from MySQL like https://github.com/prestodb/presto/blob/master/presto-resource-group-managers/src/main/java/com/facebook/presto/resourceGroups/db/ResourceGroupsDao.java and related code in https://github.com/prestodb/presto/tree/master/presto-resource-group-managers/src/main/java/com/facebook/presto/resourceGroups/db

Comment (piyushnarang) [reply presto-main/src/main/java/com/facebook/presto/server/extension/query/history/QueryHistorySQLStore.java]: - creating a dao object which defines our create table structure and the various select / insert queries and associated functions. 
- creating a mysql dao provider which instantiates it

File: presto-main/src/main/java/com/facebook/presto/server/extension/query/history/QueryHistorySQLStore.java [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+            "id bigint unsigned not null auto_increment primary key, " +
+            "cluster varchar(10) not null, " +
+            "query_id varchar(100) not null, " +
+            "query_state varchar(10) not null, " +
+            "user varchar(50) not null, " +
+            "source varchar(50), " +
+            "catalog varchar(20), " +
+            "create_time timestamp not null, " +
+            "end_time timestamp, " +
+            "query varchar(2000) not null, " +
```
Comment (piyushnarang) [inline presto-main/src/main/java/com/facebook/presto/server/extension/query/history/QueryHistorySQLStore.java]: is this too short? maybe make this long text too?

File: presto-main/src/main/java/com/facebook/presto/server/extension/query/history/QueryHistorySQLStore.java [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+            " catalog, create_time, end_time, query, query_info) values(?, ?, ?, ?, ?, ?, ?, ?, ?, ?)";
+
+    private Properties config;
+    private DataSource dataSource;
+
+    @Override
+    public void init(Properties props)
+    {
+        config = props;
+        HikariConfig dsConf = new HikariConfig(config);
```
Comment (piyushnarang) [inline presto-main/src/main/java/com/facebook/presto/server/extension/query/history/QueryHistorySQLStore.java]: if we can use the MysqlDataSource like the resource management code, we could drop this Hikari CP code.

File: presto-main/src/main/java/com/facebook/presto/server/extension/query/history/QueryHistorySQLStore.java [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+
+    @Override
+    public void close() throws IOException
+    {
+        if (dataSource instanceof Closeable) {
+            ((Closeable) dataSource).close();
+        }
+    }
+
+    private static String getCluster()
```
Comment (piyushnarang) [inline presto-main/src/main/java/com/facebook/presto/server/extension/query/history/QueryHistorySQLStore.java]: hmm I'm wondering if we can make this completely configurable. Write it out to a config file with chef so this code doesn't know about the CRITEO_ENV / DC and just gets a cluster that it needs to talk to

**Review by piyushnarang — COMMENTED** [review]
How about also adding some H2 based tests which test the QueryhistoryStore behavior?

**Review by szehon-ho — COMMENTED** [review]
looks cool for a first draft to me, just some questions

File: presto-main/src/main/java/com/facebook/presto/server/extension/query/history/QueryHistorySQLStore.java (hunk 7 above)
Comment (qinghui-xu) [reply presto-main/src/main/java/com/facebook/presto/server/extension/query/history/QueryHistorySQLStore.java]: longtext is a blob, which means the object itself is kept separately from the table, and the table only has a pointer to the data. so it might be less efficient if we want to search on a longtext field than a varchar field (afaik, we could not build index on a longtext). in the mean time, query text field is likely to be a searchable field, so it would be more relevant to be a varchar.
but maybe we could use a larger varchar field, like varchar(5000). and we can consider to truncate the query if it is too large.

File: presto-main/src/main/java/com/facebook/presto/server/extension/query/history/QueryHistorySQLStore.java (hunk 8 above)
Comment (qinghui-xu) [reply presto-main/src/main/java/com/facebook/presto/server/extension/query/history/QueryHistorySQLStore.java]: We could do that later.
For now, the hikari provides us agility to switch jdbc driver very easily (just to change `jdbcUrl` in the properties and to drop a relevant jdbc library in the classpath)

File: presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java (hunk 2 above)
Comment (qinghui-xu) [reply presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java:232]: Not in the critical path, if you mean if this will block the query execution. In fact it is called in a house keeping thread that is removing expired queries.
Also, the call to DB is done in a way that it will handle all SQL exceptions and should never fail.

**Comment by szehon-ho** [comment]
OK great, it looks good on my side to merge if Piyush gives a go

File: presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java (hunk 5 above)
Comment (qinghui-xu) [reply presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java:202]: As per the behavior of the original method `QueryTracker.getQuery`, it should throw `NoSuchElementException`.

File: presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java (hunk 6 above)
Comment (qinghui-xu) [reply presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java:210]: Same here, it is following the pattern of `getQuery`

File: presto-main/src/main/java/com/facebook/presto/server/extension/query/history/QueryHistorySQLStore.java (hunk 1 above)
Comment (qinghui-xu) [reply presto-main/src/main/java/com/facebook/presto/server/extension/query/history/QueryHistorySQLStore.java]: - Use a DAO layer to abstract all the sql out from the logic
- As currently we use Hikari to provide jdbc connection pool (thus the `DataSource`), using mysql (or another sql implementation) is just a configuration (`jdbcUrl`) to change (via some properties file), so we don't need a specific dao provider.

File: presto-main/src/main/java/com/facebook/presto/server/extension/query/history/QueryHistorySQLStore.java (hunk 9 above)
Comment (qinghui-xu) [reply presto-main/src/main/java/com/facebook/presto/server/extension/query/history/QueryHistorySQLStore.java]: Yes, will read the information directly from the extension properties file.

File: presto-main/src/main/java/com/facebook/presto/server/extension/query/history/QueryHistorySQLStore.java (hunk 1 above)
Comment (qinghui-xu) [reply presto-main/src/main/java/com/facebook/presto/server/extension/query/history/QueryHistorySQLStore.java]: It's used in our test, but as we have a DAO layer now, I'll put this constant to the DAO.

File: presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java (hunk 4 above)
Comment (qinghui-xu) [reply presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java]: Yes, we can simplify the configurations.
Now we put the implementation config into the properties file, so the extension creation relies solely on the existence of the properties file (and the awareness of the file).

File: presto-history-extension/README.md [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+* Create a properties file in presto coordinator's the file system (e.g. `${PRESTO_ROOT}/etc/extension/history-extension.properties`)
+* In the properties file, add the extension implementation property `com.facebook.presto.server.extension.query.history.QueryHistoryStore.impl` in the property file (e.g. `com.facebook.presto.server.extension.query.history.QueryHistoryStore.impl = com.facebook.presto.server.extension.query.history.QueryHistorySQLStore`). This is the implementation that we will use for the history extension
+* In the properties file, add all the necessary jdbc properties under the namespace `sql.`, especially, `sql.jdbcUrl`, that's used to determine both jdbc connection and jdbc driver (e.g. `sql.jdbcUrl = jdbc:mariadb:://localhost:3306/DB?user=root&password=myPassword`)
+* In the properties file, define appropriate `presto.env` and `presto.dc` (or define directly `presto.cluster`) which will be used to filter the origin of the query
+* Add the jvm property `extension.query.history.config` to point to the properties file before (re)starting presto coordinator (e.g. `-Dextension.query.history.config = ${PRESTO_ROOT}/etc/extension/history-extension.properties`)
+
```
Comment (piyushnarang) [inline presto-history-extension/README.md:14]: wdyt of putting a full sample properties file at the bottom?

File: presto-history-extension/pom.xml [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+        <groupId>com.facebook.presto</groupId>
+        <version>0.213</version>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>presto-history-extension</artifactId>
+
+    <properties>
+        <air.main.basedir>${project.parent.basedir}</air.main.basedir>
+        <air.check.fail-duplicate-finder>false</air.check.fail-duplicate-finder>
```
Comment (piyushnarang) [inline presto-history-extension/pom.xml:16]: remove?

File: presto-history-extension/src/main/java/com/facebook/presto/server/extension/query/history/QueryHistoryDAO.java [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+import org.jdbi.v3.sqlobject.customizer.BindBean;
+import org.jdbi.v3.sqlobject.statement.SqlQuery;
+import org.jdbi.v3.sqlobject.statement.SqlUpdate;
+import org.jdbi.v3.sqlobject.statement.UseRowMapper;
+import org.jdbi.v3.sqlobject.transaction.Transaction;
+
+public interface QueryHistoryDAO
+{
+    // DDL
+    String CREATE_TABLE = "create table query_history (" +
```
Comment (piyushnarang) [inline presto-history-extension/src/main/java/com/facebook/presto/server/extension/query/history/QueryHistoryDAO.java]: why not wrap this in a `@SQLUpdate("create table query_history(...)")` + createQueryHistoryTable() ?

File: presto-history-extension/src/main/java/com/facebook/presto/server/extension/query/history/QueryHistoryDAO.java [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+            "user varchar(50) not null, " +
+            "source varchar(50), " +
+            "catalog varchar(20), " +
+            "create_time timestamp not null, " +
+            "end_time timestamp, " +
+            "query varchar(2000) not null, " +
+            "query_info longtext not null);";
+
+    @Transaction(TransactionIsolationLevel.READ_COMMITTED)
+    @SqlUpdate("insert into query_history" +
```
Comment (piyushnarang) [inline presto-history-extension/src/main/java/com/facebook/presto/server/extension/query/history/QueryHistoryDAO.java]: wdyt of uppercasing the SQL keywords e.g - INSERT INTO query_history .. ?

File: presto-history-extension/src/main/java/com/facebook/presto/server/extension/query/history/QueryHistorySQLStore.java [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+    }
+
+    /**
+     * Find the presto cluster location, first try to use "presto.cluster", then fall back to "presto.env" - "prest.dc"
+     *
+     * @return
+     */
+    private String getCluster()
+    {
+        return config.getProperty("presto.cluster", config.getProperty("presto.env", "preprod") + "-" + config.getProperty("presto.dc", "pa4"));
```
Comment (piyushnarang) [inline presto-history-extension/src/main/java/com/facebook/presto/server/extension/query/history/QueryHistorySQLStore.java]: lets drop the fall back. It makes this fragile. If we need the cluster name / details lets mandate that. Else if we change some details elsewhere it'll come back and bite us herre.

File: presto-history-extension/src/test/java/com/facebook/presto/server/extension/query/history/TestQueryHistorySQLStore.java [hunk 15]
```diff
@@ -0,0 +1,215 @@
...
+    private void verifyQueryStats(QueryStats actual, QueryStats expected)
+    {
+        assertEquals(actual.getOperatorSummaries().size(), expected.getOperatorSummaries().size());
+        assertEquals(actual.getCreateTime(), expected.getCreateTime());
+        assertEquals(actual.getEndTime(), expected.getEndTime());
+    }
+
+    // This test is disabled because we cannot deserialize a json got from presto ui.
+    // It fails when trying to parse an empty set.
+    @Test(enabled = false)
```
Comment (piyushnarang) [inline presto-history-extension/src/test/java/com/facebook/presto/server/extension/query/history/TestQueryHistorySQLStore.java]: if you think we need to get this working then lets leave this test around. Else we can drop it I think

File: presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java [hunk 16]
```diff
@@ -46,10 +52,12 @@
 import static java.util.Objects.requireNonNull;
 
 @ThreadSafe
-public class QueryTracker<T extends TrackedQuery>
+public class QueryTracker<T extends QueryExecution>
 {
     private static final Logger log = Logger.get(QueryTracker.class);
 
+    public static final String QUERY_HISTORY_EXTENSION_CONFIG = "extension.query.history.config";
```
Comment (piyushnarang) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java]: the convention followed by the rest of the code is - a .json file and a .properties file (or just a .properties file). 
The prop file has key values (e.g. `config.properties`). Alternatively it points to a json file (e.g. access-control - https://prestosql.io/docs/current/security/built-in-system-access-control.html). 

Can we restructure this the same way? I think we could just have a .properties file, which if it exists we know we need query history. Else we dont. If it exists, we can load appropriate props. We don't need this QUERY_HISTORY_EXTENSION_CONFIG system property.

File: presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java [hunk 17]
```diff
@@ -71,6 +81,35 @@ public QueryTracker(QueryManagerConfig queryManagerConfig, ScheduledExecutorServ
...
+        Properties extensionProps;
+        try {
+            extensionProps = getExtensionConf(extensionConfigFile);
+        }
+        catch (IOException e) {
+            log.warn("Failed to load query extension config from " + extensionConfigFile);
+            return Optional.empty();
+        }
+        // The implementation class is defined as a property `com.facebook.presto.server.extension.query.history.QueryHistoryStore.impl`.
+        String extensionImplClass = extensionProps.getProperty(QueryHistoryStore.class.getName() + ".impl");
```
Comment (piyushnarang) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java:103]: an alternative (nicer?) way this is done for the access control config is they maintain a factory which maps from a string -> AccessControlFactory (`file` -> FileBasedSystemAccessControl, `allow-all` -> AllowAllSystemAccessControl, ...). We could define a name to QueryHistoryStore mapping and just have the name listed in the properties file. 
See: https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/security/AccessControlManager.java

File: presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java [hunk 18]
```diff
@@ -147,6 +186,22 @@ public T getQuery(QueryId queryId)
                 .orElseThrow(NoSuchElementException::new);
     }
 
+    public BasicQueryInfo getCurrentOrPastBasicQueryInfo(QueryId queryId)
+    {
+        return tryGetQuery(queryId).map(QueryExecution::getBasicQueryInfo)
```
Comment (piyushnarang) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java:199]: we also need to handle getBasicQueryInfo returning null right?

File: presto-history-extension/pom.xml (hunk 11 above)
Comment (qinghui-xu) [reply presto-history-extension/pom.xml:16]: The build will fail without this: there are some files with the same name introduced in the dependencies, such as
```
[WARNING] Found duplicate and different classes in [commons-logging:commons-logging:1.2, org.slf4j:jcl-over-slf4j:1.7.25]:
[WARNING]   org.apache.commons.logging.Log
[WARNING]   org.apache.commons.logging.LogConfigurationException
[WARNING]   org.apache.commons.logging.LogFactory
[WARNING]   org.apache.commons.logging.impl.NoOpLog
[WARNING]   org.apache.commons.logging.impl.SimpleLog
[WARNING] Found duplicate and different resources in [ch.vorburger.mariaDB4j:mariaDB4j-core:2.2.3, ch.vorburger.mariaDB4j:mariaDB4j-db-linux64:10.1.13, ch.vorburger.mariaDB4j:mariaDB4j-db-mac64:10.1.9, ch.vorburger.mariaDB4j:mariaDB4j-db-win32:10.1.20, ch.vorburger.mariaDB4j:mariaDB4j:2.2.3]:
[WARNING]   CHANGES.md
[WARNING]   CONTRIBUTING.md
[WARNING]   CONTRIBUTORS.md
[WARNING]   README.md
```

File: presto-history-extension/README.md (hunk 10 above)
Comment (qinghui-xu) [reply presto-history-extension/README.md:14]: Done

File: presto-history-extension/src/main/java/com/facebook/presto/server/extension/query/history/QueryHistoryDAO.java (hunk 12 above)
Comment (qinghui-xu) [reply presto-history-extension/src/main/java/com/facebook/presto/server/extension/query/history/QueryHistoryDAO.java]: Done

File: presto-history-extension/src/main/java/com/facebook/presto/server/extension/query/history/QueryHistoryDAO.java (hunk 13 above)
Comment (qinghui-xu) [reply presto-history-extension/src/main/java/com/facebook/presto/server/extension/query/history/QueryHistoryDAO.java]: Done

File: presto-history-extension/src/main/java/com/facebook/presto/server/extension/query/history/QueryHistorySQLStore.java (hunk 14 above)
Comment (qinghui-xu) [reply presto-history-extension/src/main/java/com/facebook/presto/server/extension/query/history/QueryHistorySQLStore.java]: Done. Keep only one entry `presto.cluster` in the properties file.

File: presto-history-extension/src/test/java/com/facebook/presto/server/extension/query/history/TestQueryHistorySQLStore.java (hunk 15 above)
Comment (qinghui-xu) [reply presto-history-extension/src/test/java/com/facebook/presto/server/extension/query/history/TestQueryHistorySQLStore.java]: Removed the test.

File: presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java (hunk 18 above)
Comment (qinghui-xu) [reply presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java:199]: It seems to me `QueryExecution::getBasicQueryInfo` should never return `null`.

File: presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java (hunk 5 above)
Comment (qinghui-xu) [reply presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java:202]: In fact, the REST API `QueryResource` expects `NoSuchElementException` when there is no data found.

File: presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java (hunk 17 above)
Comment (qinghui-xu) [reply presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java:103]: I think these are different scenarios.
- The AccessControlFactory has enumerable implementations, thus all of them could be registered this way.
- In our case, I intend to make this `QueryHistoryStore` a true extension, meaning that any user can provide their customized implementations (maybe based on elasticsearch, cassandra, or other stuff) to fit their needs/tech stack. So we will have to "discover" the implementation in runtime, specified by the user. The way to do it here, we rely on some parameter in the properties file.

File: presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java (hunk 16 above)
Comment (qinghui-xu) [reply presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java]: I thought about having some flexibility to define which properties file to use in runtime.
But yes, it would maybe make configuring the coordinator easier if we rely on a pre-determined file.

File: presto-main/src/main/java/com/facebook/presto/server/extension/query/history/QueryHistorySQLStore.java (hunk 7 above)
Comment (piyushnarang) [reply presto-main/src/main/java/com/facebook/presto/server/extension/query/history/QueryHistorySQLStore.java]: Think it's risky to keep this short as we have often seen fairly large queries (quite often exceeding 10K chars). I'd suggest keeping as a longtext. While searching on the query text might happen occasionally, users would be far more cranky if we truncated the query text and they were not able to see the full query when they pulled up the query history page for a query being served from the DB.

File: presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java (hunk 17 above)
Comment (piyushnarang) [reply presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java:103]: yeah the way it is done in the rest of the code e.g. for resource management is for there to be a given implementation for a given keyword. E.g. in case of resource management you have either file based / mysql based. So if you specify 'file' in you resource management config, it will pick the correct file based resource manager - https://prestodb.github.io/docs/current/admin/resource-groups.html
Feel like its cleaner to specify it that way rather than listing out the class name as a property in the properties file. That's a bit fragile, if in a refactor we move / yank out the class and roll out a release, it causes existing clusters to break unless they took care to update the property class too (which they might miss).

File: presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java (hunk 18 above)
Comment (piyushnarang) [reply presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java:199]: hmm so currently in QueryHistoryStore you have:
```
public BasicQueryInfo getBasicQueryInfo(QueryId queryId)
    {
        QueryInfo fullQueryInfo = getFullQueryInfo(queryId);
        if (fullQueryInfo == null) {
            return null;
        }
```
Do you know under what scenario full query info would return null? I see that if the query id isn't in the DB you get an exception. Or should we update the getBasicQueryInfo code to assume fullQueryInfo is never null?

File: presto-history-extension/README.md [hunk 19]
```diff
@@ -0,0 +1,29 @@
+# Presto Query History Extension
+
+This module provides an implementation for query history extension, relying on a RDBMS to persist completed queries.
+
+## How to use it
+
+To enable this extension implementation for query history, you should:
+
+* Add the jar of the extension implementation (`com.facebook.presto:presto-history-extension`) into the presto coordinator classpath.
+* Add the jar of the appropriate sql jdbc (e.g. `org.mariadb.jdbc:mariadb-java-client`) into the presto coordinator classpath.
```
Comment (piyushnarang) [inline presto-history-extension/README.md]: @qinghui-xu / @szehonCriteo - do you guys know if the default mysql client drivers will work? We already have them included in Presto's existing code so I'm wondering if we can just use them. Less moving parts and one less thing to deploy.

File: presto-history-extension/README.md [hunk 20]
```diff
@@ -0,0 +1,29 @@
...
+## Behind the scenes
+
+`QueryTracker` will look up `${PRESTO_ROOT}/etc/extension/query-history-store.properties` to create a query history store. In the properties file, it discover which implementation to use for `QueryHistoryStore`. In the case of the implementation `com.facebook.presto.server.extension.query.history.QueryHistorySQLStore`, it uses the section `sql.` to find all jdbc connection configurations.
+
+If the properties file is not present or it failed to create the extension implementation (either some necessary key-value pair is missing, or any error during the instantiation such as `ClassNotFoundException` or `SQLException`), no extension will be used.
+
+## Sample of the extension properties file
+
+```
+# cat etc/query-history-store.properties
```
Comment (piyushnarang) [inline presto-history-extension/README.md]: we can remove this line and just highlight the contents of the file

File: presto-main/etc/extension/query-history-store.properties [hunk 21]
```diff
@@ -0,0 +1,3 @@
+# com.facebook.presto.server.extension.query.history.QueryHistoryStore.impl = com.facebook.presto.server.extension.query.history.QueryHistorySQLStore
```
Comment (piyushnarang) [inline presto-main/etc/extension/query-history-store.properties]: is this file needed? if yes, can we uncomment?

File: presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java (hunk 18 above)
Comment (piyushnarang) [reply presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java]: currently the code in QueryHistorySQLStore.getBasicQueryInfo can return null:
```
public BasicQueryInfo getBasicQueryInfo(QueryId queryId)
    {
        QueryInfo fullQueryInfo = getFullQueryInfo(queryId);
        if (fullQueryInfo == null) {
            return null;
        }```
I'm not sure under what circumstances getFullQueryInfo would return null though (as you've already handled missing elements).

Comment (qinghui-xu) [reply presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java:199]: Not sure for the db driver behavior when there is no line returned for a select statement. But here it makes sense to throw directly a NoSuchElementException which is expected by the REST API in such case.

File: presto-history-extension/README.md (hunk 19 above)
Comment (qinghui-xu) [reply presto-history-extension/README.md]: The mysql driver should work with a MariaDB, I just tested. So I would replace MariaDB driver by Mysql driver, as the latter one is already bundled with presto, and we will have one dependency less.

File: presto-main/etc/extension/query-history-store.properties (hunk 21 above)
Comment (qinghui-xu) [reply presto-main/etc/extension/query-history-store.properties]: This file is needed only when we enable the extension (so probably not for a dev environment). Will remove it.

File: presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java (hunk 17 above)
Comment (qinghui-xu) [reply presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java:103]: As discussed, we could have this for a future change.

File: presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java (hunk 16 above)
Comment (qinghui-xu) [reply presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java]: Now the only configuration file is `etc/extension/query-history-store.properties`, and we rely on its presence and validity to decide whether or not to create the extension.

File: presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java (hunk 18 above)
Comment (qinghui-xu) [reply presto-main/src/main/java/com/facebook/presto/execution/QueryTracker.java:199]: I'm not sure about the jdbc implementation details: whether some of them will return null when nothing is found by a `select`. So this is here to be a safe guard. But as REST API expects a NoSuchElementException in such case, I will throw an exception from here directly.

**Review by piyushnarang — APPROVED** [review]
Looks good to me. Thanks for working through the feedback and making this change!
